File: skills/on-chain-wizard/scripts/run_onchain_analysis.py
#!/usr/bin/env python3
from __future__ import annotations

import argparse
import json
import sys
import time
from pathlib import Path
from typing import Any

import requests


SUPPORTED_CHAINS = {
    "ethereum",
    "polygon",
    "bsc",
    "arbitrum",
    "optimism",
    "base",
    "avalanche",
}

DEFAULT_API_URL = "https://esraarlhpxraucslsdle.supabase.co/functions/v1/onchain-analysis"


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(description="Run On-Chain Wizard analysis via async submit/poll API")
    parser.add_argument("--contract-address", required=True, help="0x-prefixed EVM contract address")
    parser.add_argument("--chain", required=True, help="Supported chain name")
    parser.add_argument("--abi-file", help="Path to ABI JSON file for unverified contracts")
    parser.add_argument("--api-url", default=DEFAULT_API_URL, help="Override analysis API URL")
    parser.add_argument("--poll-interval", type=int, default=20, help="Poll interval in seconds")
    parser.add_argument("--max-wait", type=int, default=420, help="Max wait time in seconds")
    parser.add_argument("--timeout", type=int, default=20, help="HTTP timeout in seconds")
    return parser.parse_args()


def validate_contract(contract_address: str) -> None:
    if not (
        contract_address.startswith("0x")
        and len(contract_address) == 42
        and all(ch in "0123456789abcdefABCDEF" for ch in contract_address[2:])
    ):
        raise ValueError("contract address must be a full 0x-prefixed 40-byte hex string")


def load_abi(path: str | None) -> Any | None:
    if not path:
        return None
    raw = Path(path).read_text(encoding="utf-8")
    return json.loads(raw)


def run_analysis(
    *,
    contract_address: str,
    chain: str,
    abi: Any | None,
    api_url: str,
    poll_interval: int,
    max_wait: int,
    timeout: int,
) -> dict[str, Any]:
    payload: dict[str, Any] = {
        "contractAddress": contract_address,
        "chain": chain,
    }
    if abi is not None:
        payload["abi"] = abi

    session = requests.Session()
    submit_response = session.post(api_url, json=payload, timeout=timeout)
    submit_response.raise_for_status()
    submit_data = submit_response.json()

    poll_url = submit_data.get("pollUrl")
    if not poll_url:
        job_id = submit_data.get("jobId")
        if not job_id:
            raise ValueError("submit response missing both pollUrl and jobId")
        poll_url = f"{api_url}?jobId={job_id}"

    start = time.time()
    while True:
        poll_response = session.get(poll_url, timeout=timeout)
        poll_response.raise_for_status()
        poll_data = poll_response.json()
        status = poll_data.get("status")

        if status in {"completed", "failed"}:
            return poll_data

        if (time.time() - start) > max_wait:
            return {
                "status": "failed",
                "error": "Analysis timed out while waiting for completion",
                "jobId": poll_data.get("jobId"),
                "phase": poll_data.get("phase"),
            }

        time.sleep(poll_interval)


def main() -> int:
    args = parse_args()
    chain = args.chain.lower().strip()

    try:
        validate_contract(args.contract_address)
        if chain not in SUPPORTED_CHAINS:
            raise ValueError(
                "unsupported chain; use one of: " + ", ".join(sorted(SUPPORTED_CHAINS))
            )
        abi = load_abi(args.abi_file)
        result = run_analysis(
            contract_address=args.contract_address,
            chain=chain,
            abi=abi,
            api_url=args.api_url,
            poll_interval=args.poll_interval,
            max_wait=args.max_wait,
            timeout=args.timeout,
        )
    except Exception as exc:
        print(json.dumps({"status": "failed", "error": str(exc)}), file=sys.stderr)
        return 1

    print(json.dumps(result, indent=2))
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: skills/on-chain-wizard/scripts/run_onchain_analysis.mjs
import process from "node:process";

const SUPPORTED_CHAINS = new Set([
  "ethereum",
  "polygon",
  "bsc",
  "arbitrum",
  "optimism",
  "base",
  "avalanche"
]);

export async function runOnchainAnalysis({
  contractAddress,
  chain,
  abi = null,
  apiUrl = "https://esraarlhpxraucslsdle.supabase.co/functions/v1/onchain-analysis",
  pollIntervalSeconds = 20,
  maxWaitSeconds = 420,
  timeoutSeconds = 20
}) {
  if (!contractAddress?.startsWith("0x") || contractAddress.length !== 42) {
    throw new Error("contract address must be a full 0x-prefixed 40-byte hex string");
  }
  if (!SUPPORTED_CHAINS.has(chain)) {
    throw new Error(`unsupported chain; use one of: ${Array.from(SUPPORTED_CHAINS).sort().join(", ")}`);
  }

  const controller = new AbortController();
  const timer = setTimeout(() => controller.abort(), timeoutSeconds * 1000);
  try {
    const payload = {
      contractAddress,
      chain
    };
    if (abi) {
      payload.abi = abi;
    }
    const submit = await fetch(apiUrl, {
      method: "POST",
      headers: {
        "Content-Type": "application/json"
      },
      body: JSON.stringify(payload),
      signal: controller.signal
    });
    if (!submit.ok) {
      throw new Error(`submit failed: ${submit.status} ${await submit.text()}`);
    }
    const submitData = await submit.json();
    let pollUrl = submitData.pollUrl;
    if (!pollUrl) {
      if (!submitData.jobId) {
        throw new Error("submit response missing both pollUrl and jobId");
      }
      pollUrl = `${apiUrl}?jobId=${submitData.jobId}`;
    }

    const started = Date.now();
    while (true) {
      const pollResponse = await fetch(pollUrl, {
        method: "GET",
        signal: AbortSignal.timeout(timeoutSeconds * 1000)
      });
      if (!pollResponse.ok) {
        throw new Error(`poll failed: ${pollResponse.status} ${await pollResponse.text()}`);
      }
      const pollData = await pollResponse.json();
      if (pollData.status === "completed" || pollData.status === "failed") {
        return pollData;
      }
      if ((Date.now() - started) > (maxWaitSeconds * 1000)) {
        return {
          status: "failed",
          error: "Analysis timed out while waiting for completion",
          jobId: pollData.jobId,
          phase: pollData.phase
        };
      }
      await new Promise((resolve) => setTimeout(resolve, pollIntervalSeconds * 1000));
    }
  } finally {
    clearTimeout(timer);
  }
}

function parseArgs(argv) {
  const args = {
    contractAddress: "",
    chain: "",
    apiUrl: "https://esraarlhpxraucslsdle.supabase.co/functions/v1/onchain-analysis",
    pollIntervalSeconds: 20,
    maxWaitSeconds: 420,
    timeoutSeconds: 20,
    abi: null
  };

  for (let index = 2; index < argv.length; index += 1) {
    const token = argv[index];
    const value = argv[index + 1];
    if (token === "--contract-address") {
      args.contractAddress = value;
      index += 1;
    } else if (token === "--chain") {
      args.chain = value;
      index += 1;
    } else if (token === "--api-url") {
      args.apiUrl = value;
      index += 1;
    } else if (token === "--poll-interval") {
      args.pollIntervalSeconds = Number(value);
      index += 1;
    } else if (token === "--max-wait") {
      args.maxWaitSeconds = Number(value);
      index += 1;
    } else if (token === "--timeout") {
      args.timeoutSeconds = Number(value);
      index += 1;
    }
  }
  return args;
}

if (process.argv[1] && import.meta.url === new URL(`file://${process.argv[1]}`).href) {
  const args = parseArgs(process.argv);
  runOnchainAnalysis(args)
    .then((result) => {
      console.log(JSON.stringify(result, null, 2));
    })
    .catch((error) => {
      console.error(JSON.stringify({ status: "failed", error: error.message || String(error) }));
      process.exitCode = 1;
    });
}
